File: 2237-longest-palindrome-by-concatenating-two-letter-words/longest-palindrome-by-concatenating-two-letter-words.go
func longestPalindrome(words []string) int {
	count := [26][26]int{};

	for _, word := range words {
		a := word[0] - 'a';
		b := word[1] - 'a';
		count[a][b]++;
	}

	length := 0;
	hasCenter := false;

	for i := 0; i < 26; i++ {
		for j := 0; j < 26; j++ {
			if i == j {
				length += (count[i][i] / 2) * 4;
				if count[i][i]%2 == 1 {
					hasCenter = true;
				}
			} else if i < j {
				pairs := min(count[i][j], count[j][i]);
				length += pairs * 4;
			}
		}
	}

	if hasCenter {
		length += 2;
	}

	return length;
}

func min(a, b int) int {
	if a < b {
		return a;
	}
	return b;
}

File: 2237-longest-palindrome-by-concatenating-two-letter-words/longest-palindrome-by-concatenating-two-letter-words.java
class Solution {
    public int longestPalindrome(String[] words) {
        int[][] count = new int[26][26];

        for (String word : words) {
            int a = word.charAt(0) - 'a';
            int b = word.charAt(1) - 'a';
            count[a][b]++;
        }

        int length = 0;
        boolean hasCenter = false;

        for (int i = 0; i < 26; i++)
            for (int j = 0; j < 26; j++) {
                if (i == j) {
                    length += (count[i][i] / 2) * 4;
                    if (count[i][i] % 2 == 1)
                        hasCenter = true;
                } else if (i < j) {
                    int pairs = Math.min(count[i][j], count[j][i]);
                    length += pairs * 4;
                }
            }

        if (hasCenter)
            length += 2;

        return length;
    }
}